### PR TITLE
at_collection

### PR DESCRIPTION
**Pull request by sachins-geekyants**
**- What I did**
- Added an at_collection folder that has all the files for the collection package.
- collection package provides abstraction on the key management (https://github.com/atsign-foundation/at_client_sdk/issues/796)

**- How I did it**
- Added AtCollectionModel specs and impl.
- Added AtCollectionModelMethods mixin to provide methods on AtCollectionModel instance

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: packages/at_client/lib/src/at_collection/at_collection_spec.dart [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import 'package:at_client/src/at_collection/model/at_operation_item_status.dart';
+
+abstract class AtCollectionSpec<T extends AtCollectionModel> {
+  /// Saves the object [T] in a self key.
+  /// A unique id is generated and stored in [T.id].
+  ///
+  /// Returns true if operation was successful else returns false.
+  ///
+  /// [expiryTime] Represents the time in milliseconds beyond which the key expires.
+  Future<bool> save(T model, {int? expiryTime});
```
Comment (VJag) [inline packages/at_client/lib/src/at_collection/at_collection_spec.dart:12]: Few questions:

1. Does the save fail? If yes, how do we report it?
    - My suggestion: Let us throw whatever exceptions that atClient.put throws and document the same here.
2. Should we expose ttb also on ATCollectioonModel (ttl, ttb)
   - My suggestion: Let us make ttl and ttb available as attributes/setters on ATCollectionModel
3. Should we make void a return type for save()?
 - My suggestion: Yes, as anything that is false is reported as AtException
 - I am also in favour of returning an AtKey as a return type

File: packages/at_client/lib/src/at_collection/at_collection_spec.dart [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  ///    /// all data till now
+  ///    for(var _data in _newAtShareOperation.allData){
+  ///      print("${_data.atSign}: ${_data.status}");
+  ///    }
+  ///  });
+  ///
+  ///  // to stop further shares
+  ///  _newAtShareOperation.stop();
+  /// ```
+  AtShareOperation share(T model, List<String> atSignsList);
```
Comment (VJag) [inline packages/at_client/lib/src/at_collection/at_collection_spec.dart:88]: Should we expose ttr and ccd as optional share-related attributes?

File: packages/at_client/lib/src/at_collection/at_collection_spec.dart [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  /// [expiryTime] Represents the time in milliseconds beyond which the key expires.
+  Future<bool> save(T model, {int? expiryTime});
+
+  /// updates object [T] using [T.id] as identifier.
+  /// Throws [Exception], if [T.id] or [T.collectionName] is empty.
+  ///
+  /// [update] also takes care of updating all the associated data with [T.id] if it has been shared with mutiple atSigns.
+  ///
+  /// returns List<AtOperationItemStatus>, where AtOperationItemStatus represents update status of individual key.
+  Future<List<AtOperationItemStatus>> update(T model, {int? expiryTime});
```
Comment (VJag) [inline packages/at_client/lib/src/at_collection/at_collection_spec.dart:20]: Is it right to say that, update updates the self key and reshares it every one with whom it was previously shared?

**Comment by gkc** [comment]
Is this still a draft or is it ready for review?

**Comment by sachins-geekyants** [comment]
> Is this still a draft or is it ready for review?

at_collection is being written from scratch in this branch [feat/at_collection_impl-2.0
](https://github.com/atsign-foundation/at_client_sdk/tree/feat/at_collection_impl-2.0)

I will close this draft PR.

**Pull request closed without merging**